lib/product-import.coffee

```diff
@@ -86,17 +88,26 @@ class ProductImport
   _processBatches: (products) ->
     batchedList = _.batchList(products, 30) # max parallel elem to process
     Promise.map batchedList, (productsToProcess) =>
-      # extract all skus from master variant and variants of all jsons in the batch
-      skus = @_extractUniqueSkus(productsToProcess)
-      predicate = @_createProductFetchBySkuQueryPredicate(skus)
-      # Check predicate size by: Buffer.byteLength(predicate,'utf-8')
-      # Todo: Handle predicate if predicate size > 8kb
-      # Fetch products from product projections end point by list of skus.
-      @client.productProjections
-      .where(predicate)
-      .staged(true)
-      .all()
-      .fetch()
+      debug 'Ensuring existence of product type in memory.'
+      @_ensureProductTypesInMemory(productsToProcess)
+      .then =>
+        if @ensureEnums
+          debug 'Ensuring existence of enum keys in product type.'
+          enumUpdateActions = @_validateEnums(productsToProcess)
+          uniqueEnumUpdateActions = @_filterUniqueUpdateActions(enumUpdateActions)
+          @_updateProductType(uniqueEnumUpdateActions)
+      .then =>
+        # extract all skus from master variant and variants of all jsons in the batch
+        skus = @_extractUniqueSkus(productsToProcess)
+        predicate = @_createProductFetchBySkuQueryPredicate(skus)
+        # Check predicate size by: Buffer.byteLength(predicate,'utf-8')
+        # Todo: Handle predicate if predicate size > 8kb
+        # Fetch products from product projections end point by list of skus.
+        @client.productProjections
+        .where(predicate)
+        .staged(true)
+        .all()
```

Comment:
Hmm, I guess we could process the list of skus in batches yes.
Or better:

```coffeescript
Promise.map productsToProcess, (p) =>
  skus = @_extractUniqueSkus(p)
  # ...
```

So basically we batch each product and we just have the predicate for the skus of one product. 

@hajoeichler ?